File: fragments/test-fragments-api.py
#!/usr/bin/env python3
"""
Fragments API 测试工具
测试创建简单页面并验证 preview 功能
"""

import requests
import json
import time
import sys

BASE_URL = "http://localhost:3001"

def test_fragments_preview():
    """测试 Fragments 的 preview 功能"""

    print("=" * 60)
    print("Fragments Preview 功能测试")
    print("=" * 60)

    # 1. 测试健康检查
    print("\n[1/4] 测试服务健康状态...")
    try:
        response = requests.get(f"{BASE_URL}/api/health", timeout=5)
        print(f"✓ 服务状态: {response.status_code}")
        if response.status_code == 200:
            print(f"  响应: {response.json()}")
    except Exception as e:
        print(f"✗ 健康检查失败: {e}")
        return False

    # 2. 创建一个简单的 Python 代码片段
    print("\n[2/4] 创建 Python 代码片段...")
    code = """
print("Hello from Fragments!")
print("Testing preview functionality")

# 简单计算
result = 2 + 2
print(f"2 + 2 = {result}")
"""

    payload = {
        "code": code,
        "language": "python"
    }

    try:
        response = requests.post(
            f"{BASE_URL}/api/sandbox",
            json=payload,
            timeout=30
        )
        print(f"✓ 创建请求状态: {response.status_code}")

        if response.status_code == 200:
            result = response.json()
            print(f"  Sandbox ID: {result.get('sandboxId', 'N/A')}")
            sandbox_id = result.get('sandboxId')
        else:
            print(f"✗ 创建失败: {response.text}")
            return False

    except Exception as e:
        print(f"✗ 创建代码片段失败: {e}")
        return False

    # 3. 等待并获取执行结果
    print("\n[3/4] 等待代码执行...")
    time.sleep(3)

    try:
        response = requests.get(
            f"{BASE_URL}/api/sandbox/{sandbox_id}",
            timeout=10
        )
        print(f"✓ 获取结果状态: {response.status_code}")

        if response.status_code == 200:
            result = response.json()
            print(f"  执行状态: {result.get('status', 'N/A')}")
            print(f"  输出预览:")
            output = result.get('output', '')
            for line in output.split('\n')[:10]:  # 只显示前10行
                print(f"    {line}")
        else:
            print(f"  响应: {response.text}")

    except Exception as e:
        print(f"✗ 获取结果失败: {e}")

    # 4. 测试 HTML preview
    print("\n[4/4] 测试 HTML preview...")
    html_code = """
<!DOCTYPE html>
<html>
<head>
    <title>Fragments Test</title>
    <style>
        body {
            font-family: Arial, sans-serif;
            padding: 20px;
            background: linear-gradient(135deg, #667eea 0%, #764ba2 100%);
            color: white;
        }
        .container {
            max-width: 600px;
            margin: 0 auto;
            background: rgba(255,255,255,0.1);
            padding: 30px;
            border-radius: 10px;
        }
        h1 { margin-top: 0; }
    </style>
</head>
<body>
    <div class="container">
        <h1>Fragments Preview Test</h1>
        <p>This is a simple HTML page to test the preview functionality.</p>
        <p>Current time: <span id="time"></span></p>
    </div>
    <script>
        document.getElementById('time').textContent = new Date().toLocaleString();
    </script>
</body>
</html>
"""

    payload = {
        "code": html_code,
        "language": "html"
    }

    try:
        response = requests.post(
            f"{BASE_URL}/api/sandbox",
            json=payload,
            timeout=30
        )
        print(f"✓ HTML 创建状态: {response.status_code}")

        if response.status_code == 200:
            result = response.json()
            sandbox_id = result.get('sandboxId')
            preview_url = result.get('previewUrl', f"{BASE_URL}/preview/{sandbox_id}")
            print(f"  Sandbox ID: {sandbox_id}")
            print(f"  Preview URL: {preview_url}")

            # 尝试访问 preview
            time.sleep(2)
            preview_response = requests.get(preview_url, timeout=10)
            print(f"✓ Preview 访问状态: {preview_response.status_code}")
            if preview_response.status_code == 200:
                print(f"  Preview 内容长度: {len(preview_response.text)} 字符")
                print(f"  包含预期内容: {'Fragments Preview Test' in preview_response.text}")
        else:
            print(f"✗ HTML 创建失败: {response.text}")

    except Exception as e:
        print(f"✗ HTML preview 测试失败: {e}")

    print("\n" + "=" * 60)
    print("测试完成")
    print("=" * 60)
    return True

if __name__ == "__main__":
    success = test_fragments_preview()
    sys.exit(0 if success else 1)

File: fragments/test-fragments-api.js
/**
 * Test Fragments API - Create and Preview
 * Tests the complete flow: generate code -> execute in sandbox -> get preview URL
 */

const FRAGMENTS_URL = 'http://localhost:3001'

// Test 1: Generate a simple Python code fragment
async function testGenerateFragment() {
  console.log('\n=== Test 1: Generate Fragment ===')

  const response = await fetch(`${FRAGMENTS_URL}/api/chat`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({
      messages: [
        {
          role: 'user',
          content: 'Create a simple Python script that prints "Hello from Fragments!" and calculates 2+2'
        }
      ],
      template: 'code-interpreter-v1',
      model: 'claude-3-5-sonnet-20241022',
      config: {
        model: 'claude-3-5-sonnet-20241022',
        temperature: 0.7
      }
    })
  })

  if (!response.ok) {
    throw new Error(`Generate failed: ${response.status} ${await response.text()}`)
  }

  // Parse streaming response
  const text = await response.text()
  console.log('Raw response:', text.substring(0, 500))

  // Extract JSON from streaming format
  const lines = text.split('\n').filter(line => line.trim())
  let fragment = null

  for (const line of lines) {
    try {
      const parsed = JSON.parse(line)
      if (parsed.code) {
        fragment = parsed
      }
    } catch (e) {
      // Skip non-JSON lines
    }
  }

  if (!fragment) {
    throw new Error('No fragment generated')
  }

  console.log('✓ Fragment generated:')
  console.log('  Template:', fragment.template)
  console.log('  Code length:', fragment.code?.length || 0)
  console.log('  Code preview:', fragment.code?.substring(0, 100))

  return fragment
}

// Test 2: Execute fragment in sandbox
async function testExecuteFragment(fragment) {
  console.log('\n=== Test 2: Execute Fragment ===')

  const response = await fetch(`${FRAGMENTS_URL}/api/sandbox`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({
      fragment: fragment,
      userID: 'test-user',
      teamID: 'test-team'
    })
  })

  if (!response.ok) {
    const error = await response.text()
    throw new Error(`Execute failed: ${response.status} ${error}`)
  }

  const result = await response.json()
  console.log('✓ Execution result:')
  console.log('  Sandbox ID:', result.sbxId)
  console.log('  Template:', result.template)

  if (result.stdout) {
    console.log('  Stdout:', result.stdout)
  }
  if (result.stderr) {
    console.log('  Stderr:', result.stderr)
  }
  if (result.url) {
    console.log('  Preview URL:', result.url)
  }

  return result
}

// Test 3: Generate and execute a web app
async function testWebApp() {
  console.log('\n=== Test 3: Generate Web App ===')

  const response = await fetch(`${FRAGMENTS_URL}/api/chat`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({
      messages: [
        {
          role: 'user',
          content: 'Create a simple Next.js page that displays "Hello World" with a blue background'
        }
      ],
      template: 'nextjs-developer-dev',
      model: 'claude-3-5-sonnet-20241022',
      config: {
        model: 'claude-3-5-sonnet-20241022',
        temperature: 0.7
      }
    })
  })

  if (!response.ok) {
    throw new Error(`Generate web app failed: ${response.status}`)
  }

  const text = await response.text()
  const lines = text.split('\n').filter(line => line.trim())
  let fragment = null

  for (const line of lines) {
    try {
      const parsed = JSON.parse(line)
      if (parsed.code) {
        fragment = parsed
      }
    } catch (e) {
      // Skip non-JSON lines
    }
  }

  if (!fragment) {
    throw new Error('No web app fragment generated')
  }

  console.log('✓ Web app fragment generated')
  console.log('  Template:', fragment.template)

  // Execute in sandbox
  console.log('\n=== Executing Web App ===')
  const execResponse = await fetch(`${FRAGMENTS_URL}/api/sandbox`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({
      fragment: fragment,
      userID: 'test-user',
      teamID: 'test-team'
    })
  })

  if (!execResponse.ok) {
    const error = await execResponse.text()
    throw new Error(`Web app execution failed: ${execResponse.status} ${error}`)
  }

  const result = await execResponse.json()
  console.log('✓ Web app deployed:')
  console.log('  Sandbox ID:', result.sbxId)
  console.log('  Preview URL:', result.url)
  console.log('\n  🌐 Open this URL in browser to see the preview!')

  return result
}

// Run all tests
async function runTests() {
  console.log('🧪 Testing Fragments API\n')
  console.log('Target:', FRAGMENTS_URL)

  try {
    // Test 1: Simple Python code
    const fragment = await testGenerateFragment()
    const result = await testExecuteFragment(fragment)

    // Test 2: Web app with preview
    const webResult = await testWebApp()

    console.log('\n✅ All tests passed!')
    console.log('\n📊 Summary:')
    console.log('  - Code generation: ✓')
    console.log('  - Code execution: ✓')
    console.log('  - Web app preview: ✓')
    console.log(`  - Preview URL: ${webResult.url}`)

  } catch (error) {
    console.error('\n❌ Test failed:', error.message)
    process.exit(1)
  }
}

runTests()
